### PR TITLE
[WIP] Update .pylintrc

### PR DESCRIPTION
**Pull request by divya-mohan0209**
Added the docparams & docstyles plugins in the MASTERS section

### CODE REVIEW

**Comment by divya-mohan0209** [comment]
Fixes issue #3388, @balopat : I've enabled the docstyles & docparams plugins. Please let me know your views.

**Review by balopat — CHANGES_REQUESTED** [review]
Thanks for opening @divya-mohan0209 - I think this is not yet working. 

1.) I am surprised that all of our docstrings are compliant :D 
2.) I actually checked out your branch and "broke" intentionally one of our docstrings that should be caught by one of these checks, e.g I couldn't get the docstyle plugin to throw warnings for a docstring starting with an empty line or missing param docs, or mixed style docs (google, numpy, etc.)

I'm using `check/pylint-changed-files` for a faster feedback loop. 
Do you mind digging a bit deeper into to make sure it actually works? If find pylint sometimes a bit odd / unintuitive - your config _looks like_ it should work.

**Comment by balopat** [comment]
Oh I know what's wrong - in our pylintrc we disable all checkers first and enable them one by one - making the list of checkers explicit. We should enable the checkers within these two plugins by adding them in the enable list in .pylintrc

**Comment by balopat** [comment]
This is the number of issues: 
docstyle: 
```
   1 bad-docstring-quotes
 156 docstring-first-line-empty
```
We should definitely start with docstyle first this PR can be that. 

Then next PRs should attack docparams issues: 
```
   2 redundant-yields-doc
  10 redundant-returns-doc
  24 differing-param-doc
 170 missing-type-doc
 305 missing-raises-doc
 321 missing-yield-type-doc
 591 missing-param-doc
 708 missing-yield-doc
1613 missing-return-type-doc
4067 missing-return-doc
```
Some of these are related to `examples/*.py` and similarly `contrib/qasm_import/_*.py` - these should be excluded from the analysis - not sure how to exclude those yet from just docs checking, but I'm running ahead of myself.

**Comment by divya-mohan0209** [comment]
So, just to be clear about what is expected:

- Add the two checks in the enable list in the .pylintrc file. 
- Address the docstyle issues
- Address the docparams issues

For the second and third to-dos, would you think it worthwhile to have an automation in place?

**Comment by balopat** [comment]
> So, just to be clear about what is expected:
> 
> * Add the two checks in the enable list in the .pylintrc file.
> * Address the docstyle issues
> * Address the docparams issues
> 

For this PR
1. let's start with adding **only** the `docstyle` plugin and `docstyle` to the enable list 
2. fix all the issues.

> For the second and third to-dos, would you think it worthwhile to have an automation in place?

For the newline issues maybe you can script it...but I think it would be easier for that couple hundred to just do it manually - as it is mostly not the matter of just an extra newline but rather actually missing description. 

I realize that this sounds a bit overwhelming as it will require to look at the classes / methods and fill in the gaps for the missing pieces. Start with the easy ones and then if you are stuck with certain classes/methods that you don't know how to address, list those and I'll help out.

**Comment by divya-mohan0209** [comment]
@balopat : The Continuous Integration / Lint check is failing. I'm not sure if that's the expected result? However I have enabled it explicitly in the .pylintrc file.

**Comment by divya-mohan0209** [comment]
I'm working on fixing the docstyles errors in the list. Shall be working on them one by one. Hope there's no urgency?

**Comment by balopat** [comment]
> @balopat : The Continuous Integration / Lint check is failing. I'm not sure if that's the expected result? However I have enabled it explicitly in the .pylintrc file.

That's the expected result! If you click on the failure you'll see the linting issues listed in the logs for that job: https://github.com/quantumlib/Cirq/pull/3475/checks?check_run_id=1406666667

The goal is now to get to a green state :) 
Yes, there is no urgency on this one, take your time and thank you so much! Feel free to ask questions!

File: dev_tools/conf/.pylintrc [hunk 1]
```diff
@@ -1,4 +1,5 @@
 [MASTER]
+load-plugins=pylint.extensions.docparams,pylint.extensions.docstyle
```
Comment (dabacon) [inline dev_tools/conf/.pylintrc:2]: interestingly I learned recently that pylint has a parallel mode (multiple cpus running lint), and using it you cannot use plugins :(

Comment (balopat) [reply dev_tools/conf/.pylintrc:2]: Interesting, that sucks a bit. Our linter takes 2.4 minutes on my Mac today - and with that it's still well below the other CI tests. I think we can fit this in. If it becomes too lengthy, we can separate the two modes - quick'n'dirty and full-with-docs. The latter could run nightly and the former could be the CI version.

**Comment by balopat** [comment]
Friendly ping @divya-mohan0209 - are you still planning working on this? It is okay if no. We can close it and someone else will pick it up!

**Comment by divya-mohan0209** [comment]
Hello @balopat , I haven't yet gotten the time owing to multiple conflicts at work. Thank you for reminding me! I shall have a look at it later today & let you know.

**Comment by divya-mohan0209** [comment]
Hello @balopat due to a role transition in my current organization, I haven't been able to find time all of last week as well. I don't think it is fair to keep stretching this on account of my inabilities to complete the work. I'll be closing the PR. Thank you for the opportunity & I genuinely am sorry for not being able to complete this.

**Comment by balopat** [comment]
No worries, thanks for getting as far as you did, the conversation was valuable in itself! If you find time again and anything interesting in the repo, feel free to give it another go! Good luck with your new role!

**Pull request closed without merging**